### PR TITLE
Enable /etc/rc.local compatibility via systemd

### PR DESCRIPTION
**Pull request by ghost**
Needed to cleanup/rename my branches

To keep track of discussion started https://github.com/NixOS/nixpkgs/pull/5825

### CODE REVIEW

**Comment by peti** [comment]
Dude, why don't you force push the new code and edit the subject of the PR, if you have to? Moving the same proposal to a new PR every couple of days is a little annoying.

**Comment by arno01** [comment]
@peti well, that's what I did, I removed my branch "rclocalfeature" (which automatically closed previous PR 5825), then created a new branch "rclocalservice" and pushed it.  Here is new PR.

As I'm still git beginner, was there any better way to rename the branch without causing first PR 5825 automatically closed?

**Comment by peti** [comment]
This PR is not particularly useful in my humble opinion, because ...

1. it batches unrelated commands into one unit, so it's impossible to see their individual status codes.
2. it's impossible to specify the context in which these commands run. The documentation says "at the start of the NixOS", but this doesn't mean anything. I fear people will put code in here that depends on certain disks being mounted or certain services being started, and then they'll run into random race conditions and open issues with Nixos complaining that our code is broken.
3. I don't see why it's a good idea to write these commands to `/etc/rc.local` and have systemd execute them from there instead of putting the commands into the systemd unit directly.

:-1: from me. I'd certainly like to see point (2) addressed in the documentation of this option before I'd consider merging it. Also, the relationship to `boot.postBootCommands` should be documented (which has been pointed out before by @bjornfor and @7c6f434c).

Personally, I feel like NixOS shouldn't offer this "rc.local" hook. I'd rather add helper functions that make it easier to define one-shot units in `configuration.nix`.

**Comment by arno01** [comment]
I agree with your points @peti  , basically you just don't need it.
Well, those who don't need this feature simply don't enable it in their `/etc/nixos/configuration.nix` :-)

There are thousands of other things that I don't need, but they are there. And this is good, because I can never know when I will need something. The idea is to have an option. Yes, one can say there is always an option to write your own .nix service module, but that's not handy when time is limited.

The `/etc/rc.local` is old school, and it doesn't go with the idea of `systemd`.
However, I do use it and I do know people who need and use it. 
Also, I already was explaining my reasons here https://github.com/NixOS/nixpkgs/pull/5819#issuecomment-70404577  and @7c6f434c made his point there too.

**Comment by arno01** [comment]
@peti regarding the point (2), I think before merging you can modify the description, right?

**Comment by edolstra** [comment]
Some comments:
* There is no need to actually instantiate `/etc/rc.local`, since you can just run the script directly from the Nix store:
```
serviceConfig.ExecStart = "${pkgs.writeScript ...}";
```
* The option should be named `boot.post<Something>Commands` indicating which stage of the boot it gets started (e.g. `postNetworkCommands`, since you've ordered it after `network.target`).
* Currently the systemd service will be restarted by `nixos-rebuild` if the unit file changes. So there should be a guard to ensure it gets run only once.

**Comment by bluescreen303** [comment]
:-1: from me as well. This option will confuse new users into thinking this might be the place to get something done. 

@arno01 Yes, there are lots of options that you don't use, but these options are safe to use and are mostly the preferred way to go if you want the thing they describe. This option you are proposing would just serve 1 user, while most others agree there are a lot of drawbacks and no advantages, with the risk that other users use it in their ignorance.

How about using a simple function in your `configuration.nix` so you can use
```
let rcLocal = script: {
  description = "rc.local command";
  wantedBy = [ "multi-user.target" ];
  after = [ "network.target" ];
  serviceConfig = {
    inherit script;
    Type = "oneshot";
    RemainAfterExit = true;
  };
};
in
{ systemd.services = {
    local-foo = rcLocal ''
      if [ -e /proc/sys/foofoo ]; then
        foo bar > /proc/sys/...
      fi
    '';

    local-bar = rcLocal "bar -v";
    local-baz = rcLocal "baz -t /dev/sda";
  };
}
```
This would have the added benefit that you have a success/fail status per item, proper restart/rebuild handling and traceable logging.

The whole issue has nothing to do with `systemd` in my opinon. It's about using the power of nixos to "programmatically" describe a system. That means that things that would be tedious (repetitive/error-prone) to do properly in other distro's, are easy with nix. We don't *want* to emulate other distro's behaviour so people feel _right at home_ without any learning curve. NixOS fixes/improves many things and it *will* take some time to learn how to wield that extra power. 

As a last point, as @peti already said, please do not close this PR just to open a new, similar one. Just update this one, otherwise the conversation becomes hard to follow.

**Comment by arno01** [comment]
@bluescreen303 I agree that it has lots of drawbacks. Nobody is perfect.
The difference is that either you need to write **+20** non-obvious or just +2 very obvious for beginner lines.

I'm a beginner, I've tried to adopt your code into my `configuration.nix` file and no success. It will probably take me another 15 or more minutes to make it work, but another human won't be patient with this.

If noone finds this PR useful, then feel free to close it. I don't mind. :-)

To those who will need it, I've corrected the description and updated commit can be found here 
https://github.com/arno01/nixpkgs/commit/cb28182abb7940847f30a8c2f6bfda06e31acdc7

>*Simplicity is the ultimate sophistication. -- Leonardo da Vinci*

**Comment by bluescreen303** [comment]
@arno01: these 20 lines are only a 1-time thing. You could turn them into a nixos module (once). Then all you (and maybe others) need to do it call that for each thingy you want in rc.local.

That amounts to a 1-line-for-each-rc.local-item translation, but with the abovementioned benefits, while avoiding all mentioned pitfalls. Can't get any simpler than that :)

If you find my example non-obvious, you should probably take a look at the [nix manual](http://nixos.org/nix/manual/#ch-expression-language). All I did was take your systemd service description and turned that into a function (which can then be reused). I had the script itself (rc.local commands) be passed in as a function argument, because that's the part that varies, while the rest stays the same. This function is now a very easy way of "building an entire systemd service" for a simple singular command in 1 line. That's the power of nix. You can abstract over tedious stuff like that.

Of course - on any other distribution - I would agree with you that for some simple commands writing an entire systemd service config file is not worth it. Especially if these are all mostly the same (same requires, same service type). 

Another thing I found is that for most things I used to put in rc.local on other distro's, there's a far superior place with nixos. I used to need some user permissions on entries in /dev or correct some directory after a daemon/service started. I used rc.local for that with some clever/dirty waiting/looping, because things like udev rules got overwritten on upgrades, or upgrades would start asking questions about keeping my changes, using the default, or trying to merge them. The rc.local tricks worked most of the time but ofcourse they sometimes lead to hard-to-trace edge-cases.

With nix, there *is* an initial investment in investigating a proper solution to anything you want. But this pays off every time, because once something works, it will stay like that, even after massive upgrades or full reinstalls. 

Oh, and please update your PR instead of linking to yet another implementation :)

**Comment by arno01** [comment]
I love Nix more and more :yum: !
I actually updated my PR just few seconds right after my previous post :wink:

**Comment by lethalman** [comment]
Apart the fact that I disagree, however rc.local is supposed to fork stuff. In fact the systemd version of rc-local is forking: http://lxr.devzen.net/source/xref/systemd/units/rc-local.service.in

**Comment by arno01** [comment]
@lethalman right, I've just corrected the code of this PR. Thanks!


just FYI on `Type=oneshot`, based on http://www.freedesktop.org/software/systemd/man/systemd.service.html

```
Behavior of oneshot is similar to simple; however, it is expected that the process has to exit before systemd starts follow-up units. RemainAfterExit= is particularly useful for this type of service. This is the implied default if neither Type= or ExecStart= are specified.

...

When Type is not oneshot, only one command may and must be given. When Type=oneshot is used, zero or more commands may be specified.
```

**Comment by jagajaga** [comment]
:-1: from me too.

**Comment by peti** [comment]
I'm closing this PR because it's thoroughly stalled by controversial opinions.
